formats/ctf-text/durin.c

```diff
@@ -0,0 +1,334 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <glib.h>
+#include <babeltrace/ctf-text/durin.h>
+
+/*
+ * XXX: libdwarf often hides pointers behind typedefs, which can lead
+ * to misleading code. For instance, `Dwarf_Debug` is in fact a
+ * pointer to an opaque struct. Therefore, passing its 'value' to
+ * functions such as `dwarf_next_cu_header` will modify its members,
+ * and therefore the behaviour of other functions using the same
+ * `Dwarf_Debug`. Use extra caution.
+ */
+
+/**
+ * Advance `dwarf_info`'s internal state to point to the next compile
+ * unit (CU) header.
+ *
+ * @param dwarf_info	Dwarf_Debug instance
+ * @param cu_offset	Out parameter, offset in bytes to the new CU
+ * @returns		0 on success, -1 on failure.
+ */
+static
+int durin_cu_next_header(Dwarf_Debug *dwarf_info, Dwarf_Unsigned *cu_offset)
+{
+	int ret;
+	Dwarf_Unsigned _cu_offset;
+
+	if (!dwarf_info || !cu_offset) {
+		goto error;
+	}
+
+	ret = dwarf_next_cu_header(*dwarf_info, NULL, NULL, NULL, NULL,
+				&_cu_offset, NULL);
+	if (ret != DW_DLV_OK) {
+		goto error;
+	}
+	*cu_offset = _cu_offset;
+
+	return 0;
+
+error:
+	return -1;
+}
+
+/**
+ * Reset the internal CU offset state of `dwarf_info`, allowing to
+ * read from the first CU again.
+ *
+ * libdwarf keeps the value of the last read CU header's offset stored
+ * internally within the opaque Dwarf_Debug structure, from which it
+ * then computes the offset of the next header and fetches
+ * it. Therefore, the only apparent way to reset this offset is to
+ * iterate over all CUs until they wrap over, short of instantiating a
+ * new Dwarf_Debug structure. This function hides this manipulation.
+ *
+ * @param dwarf_info	Dwarf_Debug instance
+ * @returns		0 on success, -1 on failure.
+ */
+static
+int durin_cu_reset_internal_state(Dwarf_Debug *dwarf_info)
+{
+	int ret;
+	Dwarf_Unsigned cu_offset;
+
+	if (!dwarf_info) {
+		goto error;
+	}
+
+	do {
+		ret = dwarf_next_cu_header(*dwarf_info, NULL, NULL, NULL, NULL,
+					&cu_offset, NULL);
+	} while (ret == DW_DLV_OK);
+
+	if (ret == DW_DLV_ERROR) {
+		goto error;
+	}
+
+	return 0;
+
+error:
+	return -1;
+}
+
+struct durin_cu *durin_cu_create(Dwarf_Debug *dwarf_info)
+{
+	int ret;
+	struct durin_cu *cu;
+
+	if (!dwarf_info) {
+		goto error;
+	}
+
+	ret = durin_cu_reset_internal_state(dwarf_info);
+	if (ret) {
+		goto error;
+	}
+
+	cu = g_new0(struct durin_cu, 1);
+	if (!cu) {
+		goto error;
+	}
+	cu->dwarf_info = dwarf_info;
+	cu->offset = 0;
+
+	return cu;
+
+error:
```

Comment:
I meant setting `cu = NULL` at the declaration. Maybe we're saying the same thing?
I would not tend to have an error path if there is no need for one (e.g. to perform some clean-up).